### PR TITLE
Ast recovery

### PR DESCRIPTION
**Pull request by mk13**
Changes in this PR: 
- Error handling of dangling open/close elements that occurs at the AST level. Synthetics added as placeholders. Refer to `recover_errors_parser.dart` for sample cases.
- Added `CloseElementAst` which is linked to `ElementAst` as a field. This separates the open component and the close component - allowing for better insertion of synthetics and error recovery.
- Added CLI tester for ast level parsing, refer to `ast_cli_tester.dart`.
- Added randomized token generator and parser to check for edge cases as well as benchmarking. Could still be improved in future iteration. Refer to `test/random_generator_test/...`
- Error handling at the ExpressionAst.parse level. If error-recovery is enabled, ast will continue to be parsed even if expression parsing fails. 
- Related to above, `String value` is added as a field to any AST that holds an expressionAst field. This ensures that there is still a link to the String version of the value even if the ExpressionAst fails to parse (in which case it would be null). 

Needed changes:
- Better handle void element cases. Not all void elements are tracked by list in `parser.dart` file. 
- Improve on error recovery of NgContent and Templates(after desugaring occurs).

Future changes for better performance:
- Change Token types to enums rather than string.
- Change error accumulation to not use Exception class, but a custom lightweight class.

### CODE REVIEW

File: lib/src/ast/close_element.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  }) = _SyntheticCloseElementAst;
+
+  /// Creates a synthetic close element AST from an existing AST node.
+  factory CloseElementAst.from(
+    TemplateAst origin,
+    String name, {
+    List<WhitespaceAst> whitespaces,
+  }) = _SyntheticCloseElementAst.from;
+
+  /// Creates a new close element AST from a parsed source
```
Comment (liuming0) [inline lib/src/ast/close_element.dart]: A dot after "source".

File: lib/src/ast/close_element.dart [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    return false;
+  }
+
+  @override
+  int get hashCode => name.hashCode;
+
+  @override
+  /*=R*/ accept/*<R, C>*/(TemplateAstVisitor/*<R, C>*/ visitor, [C context]) {
+    return visitor.visitCloseElement(this, context);
```
Comment (liuming0) [inline lib/src/ast/close_element.dart]: How about use "=>"

File: lib/src/ast/close_element.dart [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  /// Whether this is a `</template>` tag and should be directly rendered.
+  bool get isEmbeddedTemplate => name == 'template';
+
+  /// Name (tag) of the close element.
+  String get name;
+
+  /// Whitespaces at the end
```
Comment (liuming0) [inline lib/src/ast/close_element.dart]: dot

File: lib/src/ast/close_element.dart [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  List<WhitespaceAst> get whitespaces;
+
+  @override
+  String toString() {
+    final buffer = new StringBuffer('$CloseElementAst </$name> { ');
```
Comment (liuming0) [inline lib/src/ast/close_element.dart]: How about use "var buffer"

File: lib/src/ast/property.dart [hunk 5]
```diff
@@ -66,6 +69,9 @@ abstract class PropertyAst implements TemplateAst {
   /// Name of the property being set.
   String get name;
 
+  /// Unquoted value being bound to property
```
Comment (liuming0) [inline lib/src/ast/property.dart]: dot

File: lib/src/parser/recursive.dart [hunk 6]
```diff
@@ -40,6 +41,30 @@ class RecursiveAstParser {
     return nodes;
   }
 
+  CloseElementAst parseCloseElement(NgToken beginToken) {
```
Comment (liuming0) [inline lib/src/parser/recursive.dart:44]: Prefer use "var" for local variable.

File: test/ast_cli_tester.dart [hunk 7]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+import 'dart:convert';
+import 'package:angular_ast/angular_ast.dart';
+
+RecoveringExceptionHandler exceptionHandler = new RecoveringExceptionHandler();
+//ThrowingExceptionHandler exceptionHandler = new ThrowingExceptionHandler();
```
Comment (liuming0) [inline test/ast_cli_tester.dart]: Delete?

File: test/ast_cli_tester.dart [hunk 8]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+import 'dart:convert';
+import 'package:angular_ast/angular_ast.dart';
+
+RecoveringExceptionHandler exceptionHandler = new RecoveringExceptionHandler();
```
Comment (liuming0) [inline test/ast_cli_tester.dart]: Is exceptionHandler going to change? If not, "final exceptionHandler" should be used.

File: test/ast_cli_tester.dart [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+//ThrowingExceptionHandler exceptionHandler = new ThrowingExceptionHandler();
+
+List<StandaloneTemplateAst> parse(String template) {
+  return const NgParser().parsePreserve(
```
Comment (liuming0) [inline test/ast_cli_tester.dart]: How about "=>"

File: test/ast_cli_tester.dart [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    template,
+    sourceUrl: '/test/parser_test.dart#inline',
+    exceptionHandler: exceptionHandler,
+  );
+}
+
+void main() {
```
Comment (liuming0) [inline test/ast_cli_tester.dart]: Quote should be consistent. Prefer single over double.
e.g. print('--------');

File: test/ast_cli_tester.dart [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    if (exceptionHandler is RecoveringExceptionHandler) {
+      final exceptionsList = exceptionHandler.exceptions;
+      if (exceptionsList.isEmpty) {
+        print("CORRECT!");
+      } else {
+        HumanizingTemplateAstVisitor visitor =
+            const HumanizingTemplateAstVisitor();
+        String fixed = ast.map((t) => t.accept(visitor)).join('');
+        print("ORGNL: " + input);
```
Comment (liuming0) [inline test/ast_cli_tester.dart]: print('ORGNL: $input');

**Review by liuming0 — APPROVED** [review]

File: lib/src/ast/close_element.dart [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+      visitor.visitCloseElement(this, context);
+
+  /// Whether this is a `</template>` tag and should be directly rendered.
+  bool get isEmbeddedTemplate => name == 'template';
+
+  /// Name (tag) of the close element.
+  String get name;
+
+  /// Whitespaces at the end.
+  List<WhitespaceAst> get whitespaces;
```
Comment (MichaelRFairhurst) [inline lib/src/ast/close_element.dart]: Once again -- do we need to track whitespaces? Is it possible to have multiple?

File: lib/src/ast/element.dart [hunk 13]
```diff
@@ -45,11 +51,11 @@ abstract class ElementAst implements StandaloneTemplateAst {
...
   factory ElementAst.parsed(
     SourceFile sourceFile,
-    NgToken beginToken,
+    NgToken openElementStart,
     NgToken nameToken,
-    int openTagEnd,
-    int closeTagStart,
-    NgToken endToken, {
+    NgToken openElementEnd, {
+    bool isVoidElement,
```
Comment (MichaelRFairhurst) [inline lib/src/ast/element.dart]: do we need to track both things? That its void and its close element? Couldn't the former be a getter which returns `closeComplement == null`?

File: lib/src/ast/element.dart [hunk 14]
```diff
@@ -97,6 +109,20 @@ abstract class ElementAst implements StandaloneTemplateAst {
   /// Whether this is a `<template>` tag and should not be directly rendered.
   bool get isEmbeddedTemplate => name == 'template';
 
+  /// Determines whether the element tag name is void element.
+  bool get isVoidElement;
+
+  /// Whether this uses the void tag end '/>'.
+  ///
+  /// By definition, a void element can use either '>' or '/>'.
+  bool get usesVoidTagEnd;
```
Comment (MichaelRFairhurst) [inline lib/src/ast/element.dart]: is it possible to get `/ >` and do we treat that the same? In other words, do we need an offset for this?

File: lib/src/ast/element.dart [hunk 15]
```diff
@@ -207,12 +236,17 @@ class ParsedElementAst extends TemplateAst with ElementAst {
     this.stars: const [],
     this.whitespaces: const [],
   })
-      : super.parsed(beginToken, endToken, sourceFile);
+      : usesVoidTagEnd = openElementEnd.type == NgTokenType.openElementEndVoid,
```
Comment (MichaelRFairhurst) [inline lib/src/ast/element.dart]: this could also be a getter rather than calculated & stored as a member

File: lib/src/ast/property.dart [hunk 16]
```diff
@@ -127,13 +133,10 @@ class ParsedPropertyAst extends TemplateAst
     this.nameToken,
     this.suffixToken, [
     this.valueToken,
+    this.expression,
```
Comment (MichaelRFairhurst) [inline lib/src/ast/property.dart:136]: +1 to this change.

I don't like that it used to fire a dart parser to create a node! I the expression being passed in

File: lib/src/parser/recursive.dart [hunk 17]
```diff
@@ -103,23 +128,29 @@ class RecursiveAstParser {
...
           beginToken,
           prefixToken,
           decoratorToken,
           suffixToken,
           valueToken,
+          expressionAst,
           equalSignToken,
         );
       } else if (prefixType == NgTokenType.propertyPrefix) {
+        ExpressionAst expressionAst =
```
Comment (MichaelRFairhurst) [inline lib/src/parser/recursive.dart:146]: again, +1 to this change! I like the organization a lot

**Review by MichaelRFairhurst — APPROVED** [review]
Nice change! Some nits but I love this change.

File: lib/src/ast/close_element.dart (hunk 12 above)
Comment (mk13) [reply lib/src/ast/close_element.dart]: Unfortunately this isn't the PR where I will be removing this - but will be coming soon shortly.

File: lib/src/ast/element.dart (hunk 13 above)
Comment (mk13) [reply lib/src/ast/element.dart]: Misread the question. I'll try this out to see if it's still viable.

File: lib/src/ast/element.dart (hunk 14 above)
Comment (mk13) [reply lib/src/ast/element.dart]: Yes that should be doable.

File: lib/src/ast/element.dart (hunk 15 above)
Comment (mk13) [reply lib/src/ast/element.dart]: Ugly 'usesVoidTagEnd' will be removed shortly.

File: lib/src/ast/close_element.dart (hunk 12 above)
Comment (matanlurey) [reply lib/src/ast/close_element.dart]: Maybe add a comment that this will be removed?

Or use `@Deprecated` already :)

**Review by matanlurey — APPROVED** [review]
Just tried it out, looks good :)

**Comment by mk13** [comment]
Whitespaces now marked as to be deprecated. 'usesVoidTag' boolean value removed and all affecting functions also modified.

**Comment by mk13** [comment]
'isVoidElement' is no longer a field variable and a getter based on closeComplement == null instead.
